### PR TITLE
accesslog: Pass destination identity from L7 proxies to access logs

### LINKED ISSUE #9558: monitor: Reply packet have their source/destination information backwards
Running a curl with L7 policy enabled generates the following output:

```
-> Request http from 1147 ([...]) to 0 ([reserved:world]), identity 47784->2, verdict Forwarded GET http://some-website.com/ => 0
-> Response http to 1147 ([...]) from 0 ([reserved:world]), identity 47784->2, verdict Forwarded GET http://some-website.com/ => 301
```

Even though the to/from 0 is shown correctly in the monitor output, the actual source and destination fields in the `LogRecord` struct are not correctly reversed.

Relevant code: https://github.com/cilium/cilium/blob/master/pkg/monitor/logrecord.go#L72

### PR DESCRIPTION
**Pull request by jrajahalme**
Always pass both the source and destination security identities that were used
in L7 proxy policy enforcement to access logs and make sure that also
destination security identity is access logged in L7 messages in all cases. This
fixes cases where destination security identity may have been access logged
as 'WORLD' even though the destination is a pod in the cluster.

This also makes sure that the correct security identities are access logged,
even in cases where endpoint's security identity may have changed between
the time of the proxy event and log handling in the Cilium agent.

4th commit fixes an old discrepancy by passing source and destination
addressing info in access logs correctly instead of relying Hubble to swap
them around for responses. This has the benefit that also cilium monitor
output has correct security identity for L7 responses.

Note that this changes cilium proxy output that used to incorrectly show
security identities as "A->B" for both requests and responses. Now these
are formatted as "A->B" and "B->A" for requests and responses,
respectively.

Final commit avoids losing Envoy access log messages in the case the
local endpoint can not be found (any more).

This is a prerequisite for #18894

Fixes: #9558
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Cilium monitor now correctly reports security identities for L7 flows.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
5.4 flake #16852

**Comment by jrajahalme** [comment]
net-next hit by flake #12511

**Comment by jrajahalme** [comment]
Only known flakes, should be good to merge.

**Review by michi-covalent — APPROVED** [review]
:heavy_check_mark:  reviewed `pkg/hubble`

**Comment by pchaigno** [comment]
AFAICS, because of the second commit, this is a bugfix and needs to be backported to v1.11, no?

**Comment by jrajahalme** [comment]
> AFAICS, because of the second commit, this is a bugfix and needs to be backported to v1.11, no?

Correct, marking for 1.11 backport. The bug is likely not severe enough to warrant backports to 1.10 or 1.9?

**Review by glibsm — APPROVED** [review]
`pkg/hubble/parser/seven` looks good.

File: pkg/proxy/logger/logger.go [hunk 1]
```diff
@@ -105,7 +99,7 @@ func NewLogRecord(endpointInfoRegistry EndpointInfoRegistry, localEndpointInfoSo
 // fillEndpointInfo fills the EndpointInfo fields using identity sent by
 // source.
 func (lr *LogRecord) fillEndpointInfo(info *accesslog.EndpointInfo, ip net.IP, secId identity.NumericIdentity) {
-	lr.endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
+	endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
```
Comment (aditighag) [inline pkg/proxy/logger/logger.go]: Does this need to be a method receiver anymore given that `endpointInfoRegistry` is moved out of `LogRecord`.

File: daemon/cmd/fqdn.go [hunk 2]
```diff
@@ -467,15 +467,22 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
...
+		addrInfo.DstIPPort = epIPPort
+		addrInfo.DstIdentity = ep.GetIdentity()
+		addrInfo.SrcIPPort = serverAddr
+		addrInfo.SrcIdentity = serverID
+	} else {
+		addrInfo.SrcIPPort = epIPPort
+		addrInfo.SrcIdentity = ep.GetIdentity()
+		addrInfo.DstIPPort = serverAddr
+		addrInfo.DstIdentity = serverID
+	}
```
Comment (aditighag) [inline daemon/cmd/fqdn.go]: Can we move this block that populates various fields earlier in the function (L440) where it gives context for how the direction is interpreted - "The observation point is always Egress, however."? We can move the ep nil check condition before L434.

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -2248,21 +2248,6 @@ func (e *Endpoint) Delete(conf DeleteConfig) []error {
 	return errs
 }
 
-// GetProxyInfoByFields returns the ID, IPv4 address, IPv6 address, labels,
-// SHA of labels, and identity of the endpoint. Returns an error if the endpoint
-// is in the process of being deleted / has been deleted.
-func (e *Endpoint) GetProxyInfoByFields() (uint64, string, string, []string, string, uint64, error) {
```
Comment (aditighag) [inline pkg/endpoint/endpoint.go:2258]: How's this change related to the commit description?

File: pkg/envoy/accesslog_server.go [hunk 4]
```diff
@@ -164,16 +162,26 @@ func logRecord(endpointInfoRegistry logger.EndpointInfoRegistry, localEndpoint l
 		})
 	}
 
-	r := logger.NewLogRecord(endpointInfoRegistry, localEndpoint, GetFlowType(pblog), pblog.IsIngress,
+	flowType := GetFlowType(pblog)
+	// Our accesslogs from Envoy do not flip the addressing for the response logs,
+	// swap them here so that the logs have the correct info.
+	// TODO (jrajahalme): Consider doing this at our Envoy filters instead?
+	var addrInfo logger.AddressingInfo
+	if flowType == accesslog.TypeResponse {
```
Comment (aditighag) [inline pkg/envoy/accesslog_server.go:176]: This warrants a comment saying how the direction is interpreted (similar to above).

File: pkg/proxy/logger/logger.go [hunk 5]
```diff
@@ -161,9 +148,7 @@ func (logTags) Addressing(i AddressingInfo) LogTag {
 			p, err := strconv.ParseUint(port, 10, 16)
 			if err == nil {
 				lr.SourceEndpoint.Port = uint16(p)
-				if lr.ObservationPoint == accesslog.Ingress {
-					lr.fillEndpointInfo(&lr.SourceEndpoint, ip, i.SrcIdentity)
-				}
+				lr.fillEndpointInfo(&lr.SourceEndpoint, ip, i.SrcIdentity)
```
Comment (aditighag) [inline pkg/proxy/logger/logger.go]: See comment @L102.

File: pkg/proxy/logger/logger.go [hunk 6]
```diff
@@ -102,69 +102,10 @@ func NewLogRecord(endpointInfoRegistry EndpointInfoRegistry, localEndpointInfoSo
...
-// leaving the proxy at egress.
-func (lr *LogRecord) fillEgressDestinationInfo(info *accesslog.EndpointInfo, ipstr string) {
-	ip := net.ParseIP(ipstr)
-	if ip != nil {
-		lr.fillEndpointInfo(info, ip)
-	}
+// fillEndpointInfo fills the EndpointInfo fields using identity sent by
+// source.
+func (lr *LogRecord) fillEndpointInfo(info *accesslog.EndpointInfo, ip net.IP, secId identity.NumericIdentity) {
+	lr.endpointInfoRegistry.FillEndpointInfo(info, ip, secId)
```
Comment (aditighag) [inline pkg/proxy/logger/logger.go]: This is again modified in the next commit. Can we pull common changes in a separate commit?

**Review by aditighag — CHANGES_REQUESTED** [review]
While the commit descriptions mention the issues that are being fixed, I didn't fully understand the implementation details. 

>Remove redundant code from pkg/fqdn and simplify pkg/proxy/logger now
that the source and destination endpoint info is filled in the same way.

Can the clean-up related changes be moved to separate commits so that reviewers know the relevant changes to review?

File: pkg/proxy/logger/logger.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/proxy/logger/logger.go]: No, will remove this function and use `endpointInfoRegistry` directly from the call sites, thanks!

File: daemon/cmd/fqdn.go (hunk 2 above)
Comment (jrajahalme) [reply daemon/cmd/fqdn.go]: Nice, this reads now:
```
	// We determine the direction based on the DNS packet. The observation
	// point is always Egress, however.
	var flowType accesslog.FlowType
	var addrInfo logger.AddressingInfo
	if msg.Response {
		flowType = accesslog.TypeResponse
		addrInfo.DstIPPort = epIPPort
		addrInfo.DstIdentity = ep.GetIdentity()
		addrInfo.SrcIPPort = serverAddr
		addrInfo.SrcIdentity = serverID
	} else {
		flowType = accesslog.TypeRequest
		addrInfo.SrcIPPort = epIPPort
		addrInfo.SrcIdentity = ep.GetIdentity()
		addrInfo.DstIPPort = serverAddr
		addrInfo.DstIdentity = serverID
	}
```

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/endpoint/endpoint.go:2258]: Dead code after `LogRecord.localEndpointInfo` was removed. I split this change to a separate commit to make this clearer.

File: pkg/envoy/accesslog_server.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/envoy/accesslog_server.go:176]: I changed the comment to this to make this clearer:
```
	// Response access logs from Envoy inherit the source/destination info from the request log
	// message. Swap source/destination info here for the response logs so that they are
	// correct.
	// TODO (jrajahalme): Consider doing this at our Envoy filters instead?
```

File: pkg/proxy/logger/logger.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/proxy/logger/logger.go]: Done :-)

File: pkg/proxy/logger/logger.go (hunk 6 above)
Comment (jrajahalme) [reply pkg/proxy/logger/logger.go]: I pulled this out from the last commit, which is now a lot more self contained, thanks!

File: pkg/proxy/logger/logger.go [hunk 7]
```diff
@@ -50,18 +50,19 @@ const (
 // LogRecord is a proxy log record based off accesslog.LogRecord.
 type LogRecord struct {
 	accesslog.LogRecord
+}
+
+var endpointInfoRegistry EndpointInfoRegistry
 
-	// endpointInfoRegistry provides access to any endpoint's information given
```
Comment (aditighag) [inline pkg/proxy/logger/logger.go:54]: The context for the variable is now lost.

**Comment by jrajahalme** [comment]
net-next fails on a test where curl works when a policy should deny it. This is not related to the change in monitor output caused by this PR.

It is also interesting that the destination pod has no policy applied to it:
```
	 1466       Disabled           Disabled          6762       k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default       fd02::8d   10.0.0.30    ready   
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                                     
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                              
	                                                            k8s:io.kubernetes.pod.namespace=default                                                                      
	                                                            k8s:zgroup=test-k8s2                                                                                         
```

Maybe this is due to the test code `defer`ring the deletion of the policy:
```
			_, err = kubectl.CiliumClusterwidePolicyAction(netpol,
				helpers.KubectlApply, helpers.HelperTimeout)
			Expect(err).Should(BeNil(), "Policy %s cannot be applied", netpol)

			defer func() {
				_, err := kubectl.CiliumClusterwidePolicyAction(netpol,
					helpers.KubectlDelete, helpers.HelperTimeout)
				Expect(err).Should(BeNil(), "Policy %s cannot be deleted", netpol)
			}()

			// Now let's apply the policy. All request should fail.
			for _, addr := range svcAddrs {
				testCurlFailFromOutside(kubectl, ni, addr, 1)
			}
```
The test fails on the 'testCurlFailFromOutside()' above. Would be easier to debug if the policy was still in place...

**Comment by jrajahalme** [comment]
ConformanceGKE fails in the second `enable hubble` step. Also likely an unrelated flake.

**Review by aditighag — APPROVED** [review]
Thanks for the revisions - looks much cleaner now. Doesn't need my follow-up review for the latest review comment.

**Comment by michi-covalent** [comment]
/test-1.23-net-next

**Comment by jrajahalme** [comment]
net-next job did not even start the tests, looks like VM boot timed out, restarting...

**Comment by jrajahalme** [comment]
/test-1.23-net-next

**Comment by michi-covalent** [comment]
/test-1.23-net-next

**Comment by michi-covalent** [comment]
/test-1.23-net-next

**Comment by pchaigno** [comment]
There's no need to retrigger the net-next job X times. The same test will always fail because the PR branch is not up-to-date. It's the same failure as https://github.com/cilium/cilium/pull/19046#issuecomment-1064979106. It's safe to ignore.

**Comment by michi-covalent** [comment]
looks like ci-aks is also failing consistently 😞

**Comment by pchaigno** [comment]
> looks like ci-aks is also failing consistently disappointed

That's because this branch is so old it doesn't have https://github.com/cilium/cilium/pull/18911 :scream_cat: I'd strongly recommend to rebase every time you push to a PR. Anyway, it looks safe to ignore to me, since this pull request is unlikely to have any effect on IPsec connectivity in Azure.

File: pkg/proxy/logger/logger.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/proxy/logger/logger.go:54]: Thanks for the review! I restored the comment, since I had to rebase to master to pick up CI fixes anyway :-)

**Comment by michi-covalent** [comment]
looks like vm provisioning faild in https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/1558/. restarting.
```
20:40:57  ==> runtime: Running provisioner: file...
20:40:57      runtime: provision/ => /tmp/
20:41:10  ==> runtime: Running provisioner: shell...
20:41:10      runtime: Running: /tmp/vagrant-shell20220314-19200-wc3aut.sh
20:41:10      runtime: updating /etc/resolv.conf
20:41:10      runtime: /tmp/provision/compile.sh: line 28: cd: /home/vagrant/go/src/github.com/cilium/cilium: Stale file handle
20:41:10  The SSH command responded with a non-zero exit status. Vagrant
20:41:10  assumes that this means the command failed. The output for this command
20:41:10  should be in the log above. Please read the output to determine what
20:41:10  went wrong.
```

**Comment by michi-covalent** [comment]
gke-stable failed because kube-dns didn't become ready: https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/7871/
```
FAIL: Kubernetes DNS did not become ready in time
```
i see other flakes with the same error message like this one https://github.com/cilium/cilium/issues/17628 i'll re-trigger the build.

**Review by joestringer — APPROVED** [review]
